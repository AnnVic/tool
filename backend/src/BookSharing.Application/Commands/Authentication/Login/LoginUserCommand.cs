using MediatR;

namespace BookSharing.Application.Commands;

public record LoginUserCommand: IRequest<string>
{
    public string UserName { get; set; }
    public string Password { get; set; }
}
