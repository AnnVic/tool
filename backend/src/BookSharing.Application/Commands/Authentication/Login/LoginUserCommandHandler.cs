using BookSharing.Contracts.Interfaces.Services;
using BookSharing.Domain.CustomExceptions;
using BookSharing.Domain.Entities;
using MediatR;
using Microsoft.AspNetCore.Identity;

namespace BookSharing.Application.Commands.Authentication.Login;

public class LoginUserCommandHandler : IRequestHandler<LoginUserCommand, string>
{
    private readonly SignInManager<User> _signInManager;
    private readonly UserManager<User> _userManager;
    private readonly ITokenService _tokenService;

    public LoginUserCommandHandler(UserManager<User> userManager, ITokenService tokenService, SignInManager<User> signInManager)
    {
        _userManager = userManager;
        _tokenService = tokenService;
        _signInManager = signInManager;
    }

    public async Task<string> Handle(LoginUserCommand request, CancellationToken cancellationToken)
    {
        var result = await _signInManager.PasswordSignInAsync(request.UserName, request.Password, false, false);
        if (!result.Succeeded)
        {
            throw new BaseException("Invalid username or password provided. Please double-check and try again.", ErrorCode.LoginInvalidCredentials);
        }

        var token = _tokenService.GenerateToken(await _userManager.FindByNameAsync(request.UserName));

        return token;
    }
}
