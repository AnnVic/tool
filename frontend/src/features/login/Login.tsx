import { useEffect } from 'react';
import { Link, useNavigate } from 'react-router-dom';
import { toast } from 'react-toastify';
import { useFormik } from 'formik';

import { loginSchema } from './loginSchema.ts';
import { useLoginMutation } from 'services/authApiService.ts';
import alertIcon from 'assets/alert.svg';
import { setCredentials } from './loginSlice.ts';
import { useAppDispatch } from 'store/storeHooks.ts';
import {
  Button,
  Error,
  Flex,
  Form,
  LoginContainer,
  LoginFormHeader,
  LoginFormWrapper,
  Spinner,
  Input,
  Label,
  ErrorIcon,
} from 'styles/index.ts';
import { BookSharingCover } from 'components/index.ts';
import { type InitialValues, initialValues } from './constants.ts';

function Login() {
  const navigate = useNavigate();
  const dispatch = useAppDispatch();
  const [login, { isLoading, isError, isSuccess, data: credentials }] =
    useLoginMutation();
  const { handleSubmit, handleChange, handleBlur, values, errors, touched } =
    useFormik({
      initialValues,
      onSubmit,
      validationSchema: loginSchema,
    });

  useEffect(() => {
    if (credentials && isSuccess) {
      toast.success('Login successful');
      dispatch(
        setCredentials({ token: credentials.token, username: values.username }),
      );
      navigate('/');
    }

    if (isError) {
      toast.error(
        'Invalid email or password provided. Please double-check and try again.',
      );
    }
  }, [credentials, isSuccess, isError, dispatch, navigate, values.username]);

  function onSubmit(values: InitialValues) {
    login({
      username: values.username,
      password: values.password,
    });
  }

  return (
    <LoginContainer>
      <BookSharingCover />
      <LoginFormWrapper>
        <div>
          <LoginFormHeader>
            <h2>Login</h2>
            <span>Don't have an account yet?</span>
            <Link to="/signup">Sign Up</Link>
          </LoginFormHeader>
          <Form onSubmit={handleSubmit}>
            <Flex direction="column">
              <Label htmlFor="username">Username</Label>
              <Input
                $isInvalid={Boolean(!values.username && touched.username)}
                name="username"
                type="username"
                value={values.username}
                placeholder="Enter username"
                onChange={handleChange}
                onBlur={handleBlur}
                data-testid="username-input"
              />
              {!values.username && touched.username && (
                <Error>
                  <ErrorIcon src={alertIcon} alt="alert icon" />
                  {errors.username}
                </Error>
              )}
            </Flex>
            <Flex direction="column">
              <Label htmlFor="password">Password</Label>
              <Input
                $isInvalid={Boolean(!values.password && touched.password)}
                name="password"
                type="password"
                value={values.password}
                placeholder="Enter password"
                onChange={handleChange}
                onBlur={handleBlur}
                data-testid="password-input"
              />
              {!values.password && touched.password && (
                <Error>
                  <ErrorIcon src={alertIcon} alt="alert icon" />
                  {errors.password}
                </Error>
              )}
            </Flex>
            <Link to="/forgot-password">Forgot password?</Link>
            {isLoading ? (
              <Spinner />
            ) : (
              <Button type="submit" data-testid="submit-button">
                Login
              </Button>
            )}
          </Form>
        </div>
      </LoginFormWrapper>
    </LoginContainer>
  );
}

export default Login;
