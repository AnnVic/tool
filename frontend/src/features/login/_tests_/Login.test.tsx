import { beforeEach, describe, expect } from 'vitest';
import { fireEvent, screen, waitFor } from '@testing-library/react';
import { renderWithProviders } from 'utils/test-utils.tsx';
import Login from 'features/login/Login.tsx';
import { toast } from 'react-toastify';

type RenderResult = {
  asFragment: () => unknown;
};

const mocks = vi.hoisted(() => {
  return {
    isError: false,
    isSuccess: false,
    data: true,
  };
});

const mockLogin = vi.hoisted(() => {
  return {
    login: vi.fn(),
  };
});

vi.mock('services/authApiService.ts', async (importOriginal) => {
  const mod = await importOriginal();
  return {
    ...(mod as { isError: boolean; isSuccess: boolean }),
    useLoginMutation: () => [
      mockLogin.login,
      {
        isLoading: false,
        ...mocks,
      },
    ],
  };
});

vi.mock('react-toastify');

describe('Login', () => {
  let login: RenderResult;

  beforeEach(() => {
    vi.resetAllMocks();
    login = renderWithProviders(<Login />);
  });

  it('Should render Login page', () => {
    const { asFragment } = login;
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should invalidate when no inputs provided ', async () => {
    const submitButton = screen.getByTestId('submit-button');
    fireEvent.click(submitButton);

    await waitFor(() => {
      const emailError = screen.getByText('Please enter your username', {
        selector: 'p',
      });
      const passError = screen.getByText('Please enter your password', {
        selector: 'p',
      });

      expect(emailError).toBeInTheDocument();
      expect(passError).toBeInTheDocument();
    });
  });

  it('Should invalidate if wrong credentials', async () => {
    const usernameInput = screen.getByTestId('username-input');
    const passwordInput = screen.getByTestId('password-input');
    const submitButton = screen.getByTestId('submit-button');

    fireEvent.change(usernameInput, { target: { value: 'username' } });
    fireEvent.change(passwordInput, { target: { value: 'password123' } });
    fireEvent.click(submitButton);

    await waitFor(() => {
      mocks.isError = true;
      expect(toast.error).toBeCalledTimes(1);
      expect(toast.error).toBeCalledWith(
        'Invalid email or password provided. Please double-check and try again.',
      );
    });
  });

  it('Should be valid credentials', async () => {
    const usernameInput = screen.getByTestId('username-input');
    const passwordInput = screen.getByTestId('password-input');
    const submitButton = screen.getByTestId('submit-button');

    fireEvent.change(usernameInput, { target: { value: 'username' } });
    fireEvent.change(passwordInput, { target: { value: 'top-secret' } });
    fireEvent.click(submitButton);

    await waitFor(() => {
      mocks.isSuccess = true;
      mocks.isError = false;

      expect(mockLogin.login).toBeCalledTimes(1);
      expect(toast.success).toBeCalledTimes(1);
      expect(toast.success).toBeCalledWith('Login successful');
    });
  });
});
