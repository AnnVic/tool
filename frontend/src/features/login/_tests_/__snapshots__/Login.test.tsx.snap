// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login > Should render Login page 1`] = `
<DocumentFragment>
  <section
    class="sc-kEjbdu fYAfYi"
  >
    <div
      class="sc-bdfCDU iUcrJN"
      data-testid="book-sharing-cover-wrapper"
    >
      <img
        class="sc-gsTDqH"
        src="/src/assets/endava-logo.png"
      />
      <h1>
        book 
        <span>
          sharing
        </span>
      </h1>
    </div>
    <div
      class="sc-crrtmM ultnU"
    >
      <div>
        <header
          class="sc-iqHXzD dfwozg"
        >
          <h2>
            Login
          </h2>
          <span>
            Don't have an account yet?
          </span>
          <a
            href="/signup"
          >
            Sign Up
          </a>
        </header>
        <form
          class="sc-gKsecS dtuRmj"
        >
          <div
            class="sc-jSguLX ddSyvC"
            direction="column"
          >
            <label
              class="sc-fubDmA kPKLyk"
              for="username"
            >
              Username
            </label>
            <input
              class="sc-pGaPU guLLSg"
              data-testid="username-input"
              name="username"
              placeholder="Enter username"
              type="username"
              value=""
            />
          </div>
          <div
            class="sc-jSguLX ddSyvC"
            direction="column"
          >
            <label
              class="sc-fubDmA kPKLyk"
              for="password"
            >
              Password
            </label>
            <input
              class="sc-pGaPU guLLSg"
              data-testid="password-input"
              name="password"
              placeholder="Enter password"
              type="password"
              value=""
            />
          </div>
          <a
            href="/forgot-password"
          >
            Forgot password?
          </a>
          <button
            class="sc-dlfmHC kFGipp"
            data-testid="submit-button"
            type="submit"
          >
            Login
          </button>
        </form>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
